### ISSUE TITLE
Small last batches for TripletSemiHardLoss

### ISSUE THREAD

**Issue by tianhuat** [tensorflow/tensorflow#46123]
I am using Tensorflow 2.3

I am using the following code for tfrecord

    dataset = tf.data.TFRecordDataset(filenames=filenames)
    dataset = dataset.shuffle(buffer_size=100000)
    # dataset handling API
    dataset = dataset.cache()
    dataset = dataset.map(parse_examples,
                          num_parallel_calls=tf.data.experimental.AUTOTUNE)
    dataset = dataset.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
    dataset = dataset.batch(50)

Say if I have 101 images, then I will have 3 batches where the last batch contains only 1 element.

This will give error if I use TripletSemiHardLoss as it requires both positive and negative examples. I would like to request a convenient way to ensure 
a. all batches have 50 elements or 
b. to ensure each batch has both positive and negative examples

**Comment by rmothukuru** [comment]
@tianhuat,
 Can you please share the complete code and the Error Message when you are using `TripletSemiHardLoss`. Thanks!

**Comment by rmothukuru** [comment]
@tianhuat,
Can you please respond to the above comment? Thanks!

**Comment by tianhuat** [comment]
For example, you can see this:
https://stackoverflow.com/questions/59857152/nan-loss-in-keras-with-triplet-loss

**Comment by aaudiber** [comment]
You can guarantee that all batches are of size 50 by passing `drop_remainder=True` to batch:

```python
dataset = dataset.batch(50, drop_remainder=True)
```

**Comment by tilakrayal** [comment]
@tianhuat,
Could you please take a look at the above comment from the developer. Triplet loss needs triplets to work (anchor image, positive image, negative image).
You need to make sure to always get enough triplets when you sample your batches.
If you sample randomly from your dataset, the number of triplets you get depends mostly on your batch_size and num_classes.

For batch_size=16 and num_classes=100 your expected value for the number of triplets is around 2.
For datasets with num_classes=2000 your expected values are:

batch_size = 16 --> 0.1 triplets
batch_size = 32 --> 0.5 triplets
batch_size = 64 --> 1.9 triplets
batch_size = 128 --> 7.8 triplets
Most of your images will not be used to compute the loss if you sample randomly.
For datasets with even more num_classes your chances to find triplets in your batches with random sampling will be close to 0.
If your batch has 0 triplets your loss will be NaN.

```
class TripletDataset(tf.data.Dataset):

    def _generator(images_per_id, nr_of_ids):
        for i in range(nr_of_ids): 
          yield ([i] * images_per_id)
    
    def __new__(cls, datasets, images_per_id):
        nr_of_ids = len(datasets)
        choice_dataset = tf.data.Dataset.from_generator( 
            cls._generator, 
            (tf.int64), 
            (tf.TensorShape([None])),
            args=(images_per_id, nr_of_ids)
        ) \
        .shuffle(nr_of_ids) \
        .unbatch()

        return tf.data.experimental.choose_from_datasets(datasets, choice_dataset)
```
